csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -478,190 +406,17 @@ private static void Trace(string msg, Reference r, Reference resolved)
-                {
-                    // Cache the adapter endpoints.
-                    _table.AddAdapterEndpoints(reference.AdapterId, proxy.IceReference.Endpoints);
-                }
-                else if (notRegistered) // If the adapter isn't registered anymore, remove it from the cache.
-                {
-                    _table.RemoveAdapterEndpoints(reference.AdapterId);
-                }
-
-                lock (_mutex)
-                {
-                    Debug.Assert(_adapterRequests.ContainsKey(reference.AdapterId));
-                    _adapterRequests.Remove(reference.AdapterId);
-                }
-            }
-            else
-            {
-                if (proxy != null && !proxy.IceReference.IsWellKnown &&
-                    proxy.IceReference.Encoding == reference.Encoding)
-                {
-                    // Cache the well-known object reference. We only add proxies to the cache which support
-                    // the requested encoding. Unlike adapter IDs, the locator might return proxies with an
-                    // incompatible encoding.
-                    _table.AddObjectReference(reference.Identity, proxy.IceReference);
-                }
-                else if (notRegistered) // If the well-known object isn't registered anymore, remove it from the cache.
-                {
-                    _table.RemoveObjectReference(reference.Identity);
-                }
-
-                lock (_mutex)
-                {
-                    Debug.Assert(_objectRequests.ContainsKey(reference.Identity));
-                    _objectRequests.Remove(reference.Identity);
-                }
-            }
-        }
-
-        private ILocatorRegistryPrx? _locatorRegistry;
-        private readonly LocatorTable _table;
-        private readonly bool _background;
-
-        private readonly Dictionary<string, Request> _adapterRequests = new Dictionary<string, Request>();
-        private readonly object _mutex = new object();
-        private readonly Dictionary<Identity, Request> _objectRequests = new Dictionary<Identity, Request>();
     }
 
     public sealed partial class Communicator
     {
-        private readonly struct LocatorKey : IEquatable<LocatorKey>
-        {
-            public LocatorKey(ILocatorPrx prx)
-            {
-                _id = prx.Identity;
-                _encoding = prx.Encoding;
-            }
-
-            public bool Equals(LocatorKey other) => _id.Equals(other._id) && _encoding.Equals(other._encoding);
-
-            public override bool Equals(object? obj) => (obj is LocatorKey other) && Equals(other);
-
-            public override int GetHashCode() => HashCode.Combine(_id, _encoding);
-
-            private readonly Identity _id;
-            private readonly Encoding _encoding;
-        }
+        private readonly bool _backgroundLocatorCacheUpdates;
+        private readonly ConcurrentDictionary<ILocatorPrx, LocatorInfo> _locatorInfoMap =
```

Comment:
I'll add comments to the description of these two classes to explain.

Reply:
I can't see a use-case for two locator infos that share the same table ... which difference between these two locator infos would make sense? Different invocation timeouts for the underlying proxy? Different connections? Different facets as all facets share the same table (which is probably an oversight)?

Reply:
Yes, as indicated in the TODOs of the locator/router info classes... we should review how we cache router infos and locator infos. I also don't think it's useful (anymore?) to cache them based on the locator/router proxy. 

I can't really remember exactly what lead to these caching decisions 15 years ago... I believe that router infos were cached based on the router proxy because Glacier2 sessions depend on the Ice connection and different proxy settings could lead to different underlying connections (because of timeouts in particular...). For locator info, I probably implemented it the same way for consistency and at the time I don't think there was AMI support so it was a lot simpler. When the AMI support was added and then fully non-blocking connection establishment, things got more complicated because of the need to use callbacks (no async/await  support at the time...). 

The intent of this PR was to move to the new code style and to fix the code to use async/await... but yes, I'm all for revisiting how we handle locators/routers and the caching of these classes. For locators, I'd rather wait until we re-design the Ice locator Slice API to add protocol/encoding support. For routers, it would be good to find a solution for the issues mentioned in the router info comment.